Create SPI to allow prohibiting certain top-level mappings

quux00 opened the pull request:
The RootObjectNamespaceValidator allows prohibiting "namespaces" or top level fields in 
Elasticsearch mappings: both "normal" index mappings and runtime mappings.

The same validator is used to check both index-time runtime mappings and query-time index mappings.

The primary target use case for this SPI is a feature in serverless to prohibit mappings like
`_project` or `_project.my_sub_field`, but the mechanism provided in this PR allows any
possible namespace detection and prevention.

quux00 committed:
Init commit: same as cps/prohibit-_project-mappings branch

quux00 committed:
Most manual tests now pass. The two that still don't work right are:

1. One case with subobjects:false doesn't work as hoped:

// this does NOT return an error - it just silently "fails" to create the _project entry
PUT test333?error_trace=true
{
   "mappings": {
     "subobjects": false,
     "properties" : {
      "_project" : {
        "type" : "object"
      }
    }
  }
}

// YAY - this one fails with correct error message
PUT test333?error_trace=true
{
  "mappings": {
    "subobjects": false,
    "properties": {
      "_project": {
        "type": "object",
        "properties": {
          "myfield": {
            "type": "keyword"
          }
        }
      }
    }
  }
}

2. Creating a runtime mapping is not detected and prevented:

PUT /rt-index
{
  "mappings": {
    "runtime": {
      "_project": {
        "type": "keyword",
        "script": {
          "source": "emit(doc['existing_field'].value + ' some additional text')"
        }
      }
    }
  }
}

quux00 committed:
Hardcoded hack to get runtime fields with _project mappings to be detected - basic manual tests now pass

quux00 committed:
Refactored validateNamespace signature so that RootObjectMapper.processField can call it without a Mapper reference

quux00 committed:
Added namespaceValidator check to IndexService.parseRuntimeMappings

This prevents _project or _project.foo mappings in query-time runtime mappings.

For example, these now fail with "Mapping rejected: [_project.foo]. No mappings of [_project] are allowed in order to avoid conflicts with project metadata tags in serverless"
GET /blogs/_search
{
  "runtime_mappings": {
    "_project": {
      "type": "keyword",
      "script": {
        "source": "emit('somevalue')"
      }
    }
  },
  "query": {
    "match_all": {}
  },
  "fields": ["_project"]
}

GET /blogs/_search
{
  "runtime_mappings": {
    "_project.foo": {
      "type": "keyword",
      "script": {
        "source": "emit('somevalue')"
      }
    }
  },
  "query": {
    "match_all": {}
  },
  "fields": ["_project.foo"]
}

javanna commented on server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java:
```diff
@@ -548,12 +548,17 @@ public final Optional<SourceKeepMode> sourceKeepMode() {
 
+    protected void validateSubField(Mapper mapper, MappingLookup mappers) {
+        mapper.validate(mappers);
+    }
```
I am not sure I follow the need for this. it currently does the same call twice? I would expect that we only need to validate in RootObjectMapper, which takes care of the top-level

javanna commented on server/src/main/java/org/elasticsearch/index/mapper/RootObjectMapper.java:
```diff
@@ -527,6 +541,11 @@ private static boolean processField(
                 Map<String, RuntimeField> fields = RuntimeField.parseRuntimeFields((Map<String, Object>) fieldNode, parserContext, true);
+                if (namespaceValidator != null) {
+                    for (String runtimeField : fields.keySet()) {
+                        namespaceValidator.validateNamespace(null, runtimeField);
```
you are doing this twice after calling parseRuntimeFields. Shall we fold that logic into parseRuntimeFields directly>

quux00 committed:
Changed runtime field validation to be done in one place: RuntimeField.parseRuntimeFields.

Removed checks from RootObjectMapper and IndexService.
This checks both query-time and index-time runtime field mappings.

quux00 commented on server/src/main/java/org/elasticsearch/index/mapper/RootObjectMapper.java:
```diff
@@ -76,11 +76,17 @@ public static class Builder extends ObjectMapper.Builder {
         protected final Map<String, RuntimeField> runtimeFields = new HashMap<>();
         protected Explicit<Boolean> dateDetection = Defaults.DATE_DETECTION;
         protected Explicit<Boolean> numericDetection = Defaults.NUMERIC_DETECTION;
+        protected RootObjectMapperNamespaceValidator namespaceValidator;
```
No, because this is in a builder class, where it gets set outside of the constructor.
It is `final`/immutable already in `RootObjectMapper` itself.

quux00 committed:
Removed redundant call to mapper.validate(mappers) in ObjectMapper.validate

quux00 committed:
Initial set of tests for index time mappings - both standard and runtime; no query time runtime mapping tests

quux00 committed:
Added query-time runtime field tests with RootObjectMapperNamespaceValidator

quux00 committed:
Initial cleanup for PR review

quux00 committed:
Cleanup for PR review - ready for next round review

javanna reviewed (commented):
solid work, I left some minor comments.

quux00 commented on server/src/test/java/org/elasticsearch/index/mapper/RootObjectMapperTests.java:
```diff
@@ -370,6 +371,247 @@ public void testEmptyType() throws Exception {
+                }
+            }""";
+
+        // TODO: this also works - what is the difference?
```
OK, thanks. I'll remove this TODO.

quux00 committed:
Got subobjects:false tests working in RootObjectMapperTests

quux00 committed:
Merge remote-tracking branch 'elastic/main' into cross-project/prohibit-_project-mappings

elasticsearchmachine commented:
Pinging @elastic/es-search-foundations (Team:Search Foundations)

quux00 committed:
Removed ServerlessRootObjectMapperNamespaceValidator

quux00 committed:
Fixed checkstyle issue

quux00 committed:
Merge remote-tracking branch 'elastic/main' into cross-project/prohibit-_project-mappings

javanna reviewed (approved):
LGTM

quux00 committed:
Improved code comments

quux00 committed:
Merge remote-tracking branch 'elastic/main' into cross-project/prohibit-_project-mappings

quux00 committed:
Merge remote-tracking branch 'elastic/main' into cross-project/prohibit-_project-mappings

quux00 committed:
Merge remote-tracking branch 'elastic/main' into cross-project/prohibit-_project-mappings
